Run Chromatic visual testing workflow on every build

hassoncs opened the pull request:
Adds a GitHub Actions workflow to run Chromatic visual snapshot tests on pushes to main and pull requests. This will help us catch UI regressions and improve our UI testing going forward!

![image](https://github.com/user-attachments/assets/a18fd94f-cfb1-4f6f-9901-ea2952075313)
![image](https://github.com/user-attachments/assets/5654e492-0b04-4c78-ad05-95c283f2b421)

The chromatic workflow does everything needed to automatically run a chromatic screenshot build:
- Checkout code with full history
- Setup pnpm and Node.js
- Install dependencies
- Build the Storybook app
- Run Chromatic using the project token secret


This PR also updates Storybook package.json to include a `build-storybook` script and adds `chromatic` script. Adds `chromatic` to turbo.json and ignores `build-storybook.log` in Storybook .gitignore.

It also adds the ability to see "both" themes in Storybook at the same time

Commit:
ci: Add Chromatic visual testing workflow

Adds a GitHub Actions workflow to run Chromatic visual regression tests on pushes to main and pull requests.

Includes steps to:
- Checkout code with full history
- Setup pnpm and Node.js
- Install dependencies
- Build the Storybook app
- Run Chromatic using the project token secret

Also updates Storybook package.json to include a `build-storybook` script and adds `chromatic` script. Adds `chromatic` to turbo.json and ignores `build-storybook.log` in Storybook .gitignore.

Commit:
feat(storybook): Add Chromatic decorator and color story

Introduces a Chromatic decorator to render stories in both light and dark themes simultaneously for visual regression testing.

Adds a new story (`Colors.stories.tsx`) to showcase VS Code theme color variables, which will be used as a baseline for Chromatic snapshots to verify theme switching functionality.

Commit:
style(storybook): Set default font styles for Storybook preview

Sets the font family, weight, and size in the Storybook preview iframe to match VS Code's default styles. This ensures a more consistent visual representation of components during development and testing.

Commit:
feat(storybook): Add 'both' theme option for Chromatic testing

Adds a 'both' theme option to the Storybook global types.
This allows the Chromatic decorator to render both light and dark themes side-by-side for visual regression testing.
Updates the ChromaticDecorator to use the global theme setting and render both themes when 'both' is selected or when running in Chromatic.
Adjusts styling for the theme containers to improve layout and visibility during testing.

Commit:
ci: Enable GitHub PR comments for Chromatic

Adds the `token` parameter to the Chromatic workflow to enable GitHub PR comments integration.

Commit:
ci: Debug Chromatic detection and workflow

Enhance Chromatic detection logic in the Storybook decorator by adding a fallback mechanism using environment variables and URL checks.

Update the Chromatic GitHub Actions workflow to explicitly set Chromatic environment variables and add a summary of the build results to the workflow summary. Enable debug logging for the Chromatic action.

Commit:
ci: Cancel old chromatic builds if new commits are pushed

RSO commented on .github/workflows/chromatic.yml:
```diff
@@ -0,0 +1,102 @@
+              uses: chromaui/action@latest
+              with:
+                  # Chromatic project token (you'll need to add this as a GitHub secret)
+                  projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
```
Would this work when an external user creates a PR? I thought that GitHub made secrets unavailable for workflows triggered from forks.

RSO replied:
The only thing I know is that we currently can't have CI builds for dependabot PRs in our private repository, because it doesn't get access to secrets, because there's an attack vector where a malicious change is pushed to an npm package, and that package reads secrets in a `postinstall` script.

That being said, it did seem like Chromatic has a public repo that's set up in exactly the same way that you have here, so maybe this case is different? 😕

hassoncs replied:
AI Summary of the research. TLDR; it's ok for Chromatic

The Security Concern is Real
RSO is correct that GitHub doesn't provide secrets to workflows triggered by external forks or dependabot PRs. This is intentional security - imagine if a malicious actor could:

Fork your repo
Add a malicious postinstall script to package.json that exfiltrates secrets
Create a PR that triggers your CI with access to your secrets
That would be a disaster.

Why Chromatic is Different
The key difference with Chromatic is in what the secret actually does:

Traditional secrets (API keys, database passwords, deployment tokens): Give access to critical infrastructure
Chromatic project token: Only allows uploading visual snapshots to your Chromatic project
The Tradeoff for Open Source
When you share a Chromatic project token publicly (as the docs suggest), you're essentially saying:

✅ "Anyone can run visual tests and consume my snapshot quota"
❌ "But they can't access my account, change settings, or approve changes"
This is why it works for open source - the "blast radius" of a compromised Chromatic token is limited to snapshot quota usage, not system compromise.

chrarnoldus replied:
PRs from external users currently require approval from us before workflows run.

![image](https://github.com/user-attachments/assets/083443cb-c846-4873-934d-1bd48b4fa144)

Commit:
ci: Skip Chromatic for external forks and dependabot PRs

Updates the Chromatic workflow to explicitly skip the Chromatic build step for pull requests originating from external forks or created by dependabot.

This prevents build failures due to lack of access to necessary secrets in external forks and avoids unnecessary Chromatic builds for automated dependency updates. A summary step is added to indicate when Chromatic testing was skipped.

Commit:
ci: Update Chromatic workflow triggers and summary details

Adjust Chromatic workflow to run on push events to main branch in addition to pull requests from internal forks.

Update the summary output to use 'N/A' for counts when the values are zero or undefined, and correct the snapshot count variable name.

Commit:
ci: Configure Chromatic project token and skip patterns

Move Chromatic project token to package.json for public access.
Update Chromatic GitHub Action to use the public token and configure skip patterns for dependabot and renovate PRs directly in the action.
Remove manual skip logic and summary steps.
Auto-accept changes on the main branch.

Commit:
Merge branch 'main' into chromatic-screenshot-testing

hassoncs commented:
Chromatic's recommendation is just to make the key public for open source projects. I figured we'd see how quickly we hit our generation limits and decide if we like that approach going forward

chrarnoldus committed:
Merge branch 'main' into chromatic-screenshot-testing
